README.md
This is my personal solution (using Java) to the technical assessment https://github.com/emilybache/GildedRose-Refactoring-Kata from a recruitment process 

Requirements from published on https://github.com/emilybache/GildedRose-Refactoring-Kata/blob/master/GildedRoseRequirements.txt 

After reading the document, and having a look to the GitHub repository I understood that I was facing a code refactoring code.

I didn't know this exercise, even though I found out to be very popular and running for over 8 years with different solutions. I had a look to the author of this repository (Emily Bache) and found out that she used this very same exercise to give a lesson about Advanced Testing and Refactoring Techniques.
The first step prior to refactoring code is creating a nest of test. Emily encourages the use of approval tests prior to start refactoring and recommends https://approvaltests.com/   I consider it could be a necessary approach (in a real life complex scenario), but in my opinion, out of scope of this assesment (I will try it later for myself)

I had much fun trying myself to solve the assessment, even though I had to take out time from my family and complete it during the weekend. I could complement some new lessons I learned by reading refactoring theory with my intuition and previous knowledge in rewriting code (something that I've been forced to do several times during my 25 years working as an IT consultant)

The exercise can be solved in many programming languages. I've chosen Java, because I find the output to be more readable, and I have already the environment installed (from a recent Spring Boot Microservices project). But I will consider repeating this exercise using other languages as Pyhon or Scala because it will help me to refresh my coding skills. I am already thinking in bringing the refactoring theory beyond the limitations of the assesment (you are not allow to modify the base class Item) to use class inheritance.

Since the original code is on GitHub, I will use a public repository in one of my accounts to publish my solution.  https://github.com/alexleidenpm/GildedRose-Refactoring
I would also follow good Git practices, creating branches for each change  (But I will not merge those branches and delete them afterwards, so you can see the steps I made in every commit). Working alone in this project, it was not also necessary to enforce pull requests.

After cloning the repository, I copied the Java dependencies (pom.xml) and java source code into a new maven's project in IntellIJ IDEA (a Java IDE I'm comfortable with) and added to my local git repository. I added target folder to .gitignore. I created a README.md file and pushed the changes to GitHub.


changes.md
07 - add conjured items

Once we have the code cleaned we can consider conjured items.

If we were allowed to modify the item class we will easily add a property there, or better we will use inheritance

With the current limitations, we can suppose conjured items have their name starting with "Conjured"

Since it's not said anything about those conjured items, we can suppose that there are outside of
the three special cases considerer before (Aged Brie, Backstage passes or Sulfuras)  

Now we have our beautiful switch which checks the name property. To fit these items in our code, one easy solution is 
adding code in the default block and check if the name starts with "Conjured" 

- updateQuality() in GildedRose class:
    - Added code in the default switch case checking if the item is conjured (its name starts with "Conjured")
    - if we have a conjured item, we need to decrement its quality twice as fast as a normal item
    - with the new logic that was not applied before, the test_results.txt file is updated
        
---

06 - join switch code blocks

If we have a look to the current code, we can see that name property is evaluated (switch) in two different parts

There is also a middle part which decrements sellin always except with item Sulfuras

- updateQuality() in GildedRose class:
    - I will first decrement the sellin variable before the switch block for all except Sulfuras.
    I could also add in every other case of the switch and avoid one comparison
    - I will then invert again the condition and extract this code into a method decrementDaysToSell(). 
    I think the code is clearer 
    - decrementing sellin at the beginning makes that the conditions for backstage passes are updated to 10 and 5 days
    This was also stated in the requirements: 
    "Quality increases by 2 when there are 10 days or less and by 3 when there are 5 days or less"
    Again, the code is easier to understand
    - To join the second switch case we have to duplicate the codition 
    "if (item.sellIn < 0)" in every case of the switch, except Sulfuras 
    
---

05 - Simplify if and replace with switch

- updateQuality() in GildedRose class:
    - group the different if conditions on item names into switch clausules 

---

04 - Extract code that update quality into methods

- updateQuality() in GildedRose class:
    - incrementItemQuality() always checks if quality<50
    - decrementItemQuality() check  quality > 0. If another condition is used we can just change their order
    - decrementItemQuality() fixed bug with automatic refactoring
    
---

03 - Simplify conditions. 

- updateQuality() in GildedRose class:
    - Backstage passes condition is always false --> remove if() block
    - Backstage passes condition is always true --> remove if() condition, keep block code 
    - 'item.quality - item.quality' expression can be simplified with 0
     
---

02 - Used positive conditions
     
Trying to find logic in updateQuality(). It's still very difficult to follow the nested conditions inside the method. Even more, when you find positive and negative conditions involved.
     
To simplify the code I will convert all negative conditions to positive
     
Additionally, I would like to separate the different cases, so I will break the first big if loop with and OR condition into two blocks
     
- updateQuality() in GildedRose class:
    - Refactor negative conditions (!item) to positive
    - split OR condition ( || ) in if block into two blocks
     
---

01 - Extract code from the main loop to a method 

I try to understand what the code does high level, without entering into the nested conditional loops details

I noticed that basically it loops through a list of items and updates its quality.

Items haven't any dependency between them, and their properties evaluated one at a time

--> updateQuality() in GildedRose class:
- Improve readability. Use item instead of items[i]
- use foreach structure
- extract code inside main loop to updateItemQuality()

---

00 - Initial code

- Changed tested days to 10.
- Added test_results.txt file 


src/test/java/com/gildedrose/TexttestFixture.java
package com.gildedrose;

public class TexttestFixture {
    public static void main(String[] args) {
        System.out.println("OMGHAI!");

        Item[] items = new Item[] {
                new Item("+5 Dexterity Vest", 10, 20), //
                new Item("Aged Brie", 2, 0), //
                new Item("Elixir of the Mongoose", 5, 7), //
                new Item("Sulfuras, Hand of Ragnaros", 0, 80), //
                new Item("Sulfuras, Hand of Ragnaros", -1, 80),
                new Item("Backstage passes to a TAFKAL80ETC concert", 15, 20),
                new Item("Backstage passes to a TAFKAL80ETC concert", 10, 49),
                new Item("Backstage passes to a TAFKAL80ETC concert", 5, 49),
                // this conjured item does not work properly yet
                new Item("Conjured Mana Cake", 3, 6) };

        GildedRose app = new GildedRose(items);

        int days = 10;
        if (args.length > 0) {
            days = Integer.parseInt(args[0]) + 1;
        }

        for (int i = 0; i < days; i++) {
            System.out.println("-------- day " + i + " --------");
            System.out.println("name, sellIn, quality");
            for (Item item : items) {
                System.out.println(item);
            }
            System.out.println();
            app.updateQuality();
        }
    }

}

src/main/java/com/gildedrose/GildedRose.java
// Revision 07 - add conjured items
package com.gildedrose;

class GildedRose {
    Item[] items;

    public GildedRose(Item[] items) {
        this.items = items;
    }

    public void updateQuality() {
        for (Item item : items ) {
            updateItemQuality(item);
        }
    }

    private void updateItemQuality(Item item) {
        decrementDaysToSell(item);
        switch (item.name) {
            case "Sulfuras, Hand of Ragnaros":
                break;

            case "Aged Brie":
                incrementItemQuality(item);
                if (item.sellIn < 0) {
                    incrementItemQuality(item);
                }
                break;

            case "Backstage passes to a TAFKAL80ETC concert":
                incrementItemQuality(item);
                if (item.sellIn < 10) {
                    incrementItemQuality(item);
                }
                if (item.sellIn < 5) {
                    incrementItemQuality(item);
                }
                if (item.sellIn < 0) {
                    item.quality = 0;
                }
                break;

            default:
                decrementItemQuality(item);
                boolean isConjured = item.name.startsWith("Conjured");
                if(isConjured)  {
                    decrementItemQuality(item);
                }
                if (item.sellIn < 0) {
                    decrementItemQuality(item);
                    if(isConjured)  {
                        decrementItemQuality(item);
                    }
                }
                break;
        }
    }

    private void decrementDaysToSell(Item item) {
        if (!item.name.equals("Sulfuras, Hand of Ragnaros"))  {
            item.sellIn = item.sellIn - 1;
        }
    }

    private void decrementItemQuality(Item item) {
        if (item.quality > 0) {
            item.quality = item.quality - 1;
        }
    }

    private void incrementItemQuality(Item item) {
        if (item.quality < 50) {
            item.quality = item.quality + 1;
        }
    }
}